habitat-lab/habitat/tasks/nav/nav.py

```diff
@@ -1289,16 +1402,553 @@ def step(
+                self.w2v_controller.velocity_feedback_control(
+                    camera_pitch_angular_err,
+                    self._acc_ang_camera_pitch,
+                    self._w_max_camera_pitch,
+                )
+            )
+
+            # Apply action and step simulation
+            next_agent_state = self._apply_camera_pitch_velocity_action(
+                camera_pitch_angular_velocity,
+                time_step=self._time_step,
+            )
+            next_agent_state = self._apply_velocity_action(
+                linear_velocity, angular_velocity, time_step=self._time_step
+            )
+            xyt = self._agent_state_to_xyt(next_agent_state)
+
+            # Complete action early if commanded speed is low
+            if (
+                abs(linear_velocity) < self._min_abs_lin_speed
+                and abs(angular_velocity) < self._min_abs_ang_speed
+                and abs(camera_pitch_angular_velocity)
+                < self._min_abs_ang_speed_camera_pitch
+            ):
+                break
+
+        return self._get_agent_observation(next_agent_state)
+
+    @staticmethod
+    def _agent_state_to_xyt(agent_state):
+        """
+        Home robot coordinate system: +z = up
+        Habitat coordinate system: +y = up
+
+        We map the coordinate systems by assuming the +x axis is shared
+        x -> x
+        y -> -z
+        rz -> -ry
+        """
+        # Convert rotation to rotation vector
+        q = agent_state.rotation
+        if type(q) is mn.Quaternion:
+            q = quaternion.from_float_array(
+                np.array([q.scalar] + list(q.vector))
+            )
+        elif type(q) is np.ndarray:
+            q = quaternion.from_float_array(q)
+        elif type(q) is not quaternion.quaternion:
+            raise TypeError("Invalid agent_state.rotation type.")
+
+        agent_rotvec = quaternion.as_rotation_vector(q)
+
+        return np.array(
+            [
+                agent_state.position[0],
+                -agent_state.position[2],
+                agent_rotvec[1] + np.pi / 2.0,
+            ]
+        )
+
+
+@registry.register_task_action
+class MoveForwardWaypointAction(WaypointAction):
+    name: str = "move_forward_waypoint"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._forward_step_size = self._config.forward_step_size
+        self._max_wait_duration = self._config.max_wait_duration
+
+    def step(self, *args: Any, **kwargs: Any):
+        r"""Update ``_metric``, this method is called from ``Env`` on each
+        ``step``.
+        """
+        xyt_waypoint = np.array([self._forward_step_size, 0.0, 0.0])
+        return self._step_rel_waypoint(
+            xyt_waypoint, self._config.max_wait_duration, *args, **kwargs
+        )
+
+    @property
+    def action_space(self):
+        return EmptySpace()
+
+
+@registry.register_task_action
+class TurnLeftWaypointAction(WaypointAction):
```

Comment:
All this classes needs doc strings.  As well as description of different action spaces and what is Waypoint.